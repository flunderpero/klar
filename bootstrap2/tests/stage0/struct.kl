struct Address:
    street str
    number i32
end

struct Person:
    name str
    age i32
    address Address -- Structs can be nested.
end

-- Struct as parameter.
fn print_person(p Person):
print(p.address.street)
    let s = f"Person: name={p.name}, age={p.age}"
    -- Read nested struct field.
    let address = p.address
    s = f"{s}, address={address.street}"
    -- Read nested struct field using dot notation.
    s = f"{s}, address={p.address.number}"
    print(s)
end

fn create_person(name str, age i32) Person:
    return Person(name, age, Address("Street", 123))
end

fn main():
    let p = create_person("John", 20, Address("Street", 123))
    print_person(p)
    -- Update struct field.
    p.name = "Jane"
    p.age = 21
    -- Update nested struct field.
    p.address.street = "New Street"
    p.address.number = 456
    print_person(p)
end
