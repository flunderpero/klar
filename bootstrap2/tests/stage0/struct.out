Street
Person: name=John, age=20, address=Street, address=123
New Street
Person: name=Jane, age=21, address=New Street, address=456
